general/chainerrl/baselines/action_wrappers.py

```diff
@@ -0,0 +1,340 @@
+"""
+MIT License
+
+Copyright (c) Preferred Networks, Inc.
+"""
+
+from collections import OrderedDict
+import copy
+
+import numpy as np
+
+
+BINARY_KEYS = ['forward', 'back', 'left', 'right', 'jump', 'sneak', 'sprint', 'attack']
+ENUM_KEYS = ['place', 'equip', 'craft', 'nearbyCraft', 'nearbySmelt']
+ALL_ORDERED_KEYS = BINARY_KEYS + ['camera'] + ENUM_KEYS
+# Number of actions for each enum elements on each environment
+NUM_ENUM_ACTIONS = {
+    'MineRLTreechop-v0': {},
+    'MineRLNavigate-v0': OrderedDict([
+        ('place', 2),
+    ]),
+    'MineRLNavigateExtreme-v0': OrderedDict([
+        ('place', 2),
+    ]),
+    'MineRLObtainIronPickaxe-v0': OrderedDict([
+        ('place', 7),
+        ('equip', 7),
+        ('craft', 5),
+        ('nearbyCraft', 8),
+        ('nearbySmelt', 3),
+    ]),
+    'MineRLObtainDiamond-v0': OrderedDict([
+        ('place', 7),
+        ('equip', 7),
+        ('craft', 5),
+        ('nearbyCraft', 8),
+        ('nearbySmelt', 3),
+    ]),
+}
+
+
+class DiscreteActionConverter:
+    '''
+    Converter from original action dict to a discrete action.
+    This converter is consisted of a sequence of queries.
+
+    Parameters
+    ----------
+    query (list of tuple): conditions of selecting each action.
+        each tuple consisted 4 elements: query_type, condition_parameter, referring element, and return
+        query_type is either less, more, or same.
+        When the value of referring element satisfies a pair of query_type and condition_parameter,
+        then the action speficied by return is taken.
+        If multiple queries are satisfied, only first one is used.
+        If none of multiple queries is satisfied. Return 0 as noop.
+    '''
+    def __init__(self, query):
+        self.query = query
+
+    def _comp(self, value_a, value_b, operator):
+        if operator == 'less':
+            return value_a < value_b
+        elif operator == 'more':
+            return value_a > value_b
+        elif operator == 'same':
+            return abs(value_a - value_b) < 1e-3
+        else:
+            raise ValueError('Unknown operator {}'.format(operator))
+
+    def __call__(self, action):
+        episode_len = len(action['forward'])
+        ret = np.zeros((episode_len, 1), dtype=np.int32)
+        for operator, comp, action_name, value in reversed(self.query):
```

Comment:
you say first one will be used, but wouldn't reversed make you choose the last one instead?

Reply:
is it a problem if you return any other query? or would any of them work?

Reply:
Is your question whether it is possible to return a satisfied value other than the first one?

Returning non-first value is OK, but we need to fix one in some way because `discrete` wrapper allows only one action per frame. This setting comes from the hypothesis that there are some kind of the priority (maybe based on the rarity) of actions and some (rare) action should prevail in this case.

Reply:
I see, then there shouldn't be any issue with the current code, besides maybe the comment in the docstring